src/topology/inseparable.lean

```diff
@@ -178,12 +179,217 @@ def specialization_preorder : preorder X :=
+  stable_under_generalization sᶜ :=
+λ x y e, mt (hs e)
+
+lemma stable_under_specialization_compl_iff {s : set X} :
+  stable_under_specialization sᶜ ↔ stable_under_generalization s :=
+⟨λ h, compl_compl s ▸ h.compl, λ h, h.compl⟩
+
+lemma stable_under_generalization_compl_iff {s : set X} :
+  stable_under_generalization sᶜ ↔ stable_under_specialization s :=
+⟨λ h, compl_compl s ▸ h.compl, λ h, h.compl⟩
+
+lemma stable_under_specialization_sUnion (S : set (set X))
+  (H : ∀ s ∈ S, stable_under_specialization s) : stable_under_specialization (⋃₀ S) :=
+λ x y e ⟨s, hs, hxs⟩, ⟨s, hs, H s hs e hxs⟩
+
+lemma stable_under_specialization_sInter (S : set (set X))
+  (H : ∀ s ∈ S, stable_under_specialization s) : stable_under_specialization (⋂₀ S) :=
+λ x y e hx s hs, H s hs e (hx s hs)
+
+lemma stable_under_generalization_sUnion (S : set (set X))
+  (H : ∀ s ∈ S, stable_under_generalization s) : stable_under_generalization (⋃₀ S) :=
+λ x y e ⟨s, hs, hxs⟩, ⟨s, hs, H s hs e hxs⟩
+
+lemma stable_under_generalization_sInter (S : set (set X))
+  (H : ∀ s ∈ S, stable_under_generalization s) : stable_under_generalization (⋂₀ S) :=
+λ x y e hx s hs, H s hs e (hx s hs)
+
+lemma stable_under_specialization_Union {ι : Sort*} (S : ι → set X)
+  (H : ∀ i, stable_under_specialization (S i)) : stable_under_specialization (⋃ i, S i) :=
+λ x y e ⟨s, ⟨i, (hi : S i = s)⟩, hxs⟩, ⟨_, ⟨i, rfl⟩, H i e (hi.symm ▸ hxs)⟩
+
+lemma stable_under_specialization_Inter {ι : Sort*} (S : ι → set X)
+  (H : ∀ i, stable_under_specialization (S i)) : stable_under_specialization (⋂ i, S i) :=
+λ x y e hx s ⟨i, (hi : S i = s)⟩, hi ▸ H i e (hi.symm ▸ hx _ ⟨i, hi⟩)
+
+lemma stable_under_generalization_Union {ι : Sort*} (S : ι → set X)
+  (H : ∀ i, stable_under_generalization (S i)) : stable_under_generalization (⋃ i, S i) :=
+λ x y e ⟨s, ⟨i, (hi : S i = s)⟩, hxs⟩, ⟨_, ⟨i, rfl⟩, H i e (hi.symm ▸ hxs)⟩
+
+lemma stable_under_generalization_Inter {ι : Sort*} (S : ι → set X)
+  (H : ∀ i, stable_under_generalization (S i)) : stable_under_generalization (⋂ i, S i) :=
+λ x y e hx s ⟨i, (hi : S i = s)⟩, hi ▸ H i e (hi.symm ▸ hx _ ⟨i, hi⟩)
+
+lemma stable_under_specialization_iff_forall_closure_subset {s : set X} :
+  stable_under_specialization s ↔ ∀ x ∈ s, closure {x} ⊆ s :=
+by simpa only [subset_def, ← specializes_iff_mem_closure, @forall_swap (_ ∈ s)]
+
+alias stable_under_specialization_iff_forall_closure_subset ↔
+  stable_under_specialization.closure_subset _
+
+lemma stable_under_specialization_iff_Union_subset {s : set X} :
+  stable_under_specialization s ↔ (⋃ x ∈ s, closure {x}) ⊆ s :=
+by simp_rw [Union_subset_iff, stable_under_specialization_iff_forall_closure_subset]
+
+lemma stable_under_specialization_iff_Union_eq {s : set X} :
+  stable_under_specialization s ↔ (⋃ x ∈ s, closure {x}) = s :=
+stable_under_specialization_iff_Union_subset.trans
+  ⟨λ h, le_antisymm h $ λ x h, ⟨_, ⟨_, rfl⟩, _, ⟨h, rfl⟩, subset_closure $ mem_singleton x⟩, eq.le⟩
```

Comment:
These can also be generalized to upper_lower, with `closure {x}` repalced by `set.Iic x`, right? And `⋃ x ∈ s, closure {x}` is `lower_closure s`, and the last two may follow from [gi_lower_closure_coe](https://leanprover-community.github.io/mathlib_docs/order/upper_lower.html#gi_lower_closure_coe).